src/modules/be-water-smart/be-water-smart.component.ts

```diff
@@ -0,0 +1,686 @@
+          algorithms[algorithm.name] = algorithm.name;
+        });
+        this["optionsAlgorithm"] = algorithms;
+        this["algorithms"] = response.algorithms;
+      },
+      error: error => {
+        console.log(error);
+      },
+    });
+  }
+
+  /**
+   * Extraction Method for VirtualMeters
+   * @param extractionMethod the function to use for the api call
+   */
+  extractVirtualMetersInRecord(
+    extractionMethod: () => Observable<AllVirtualMeters>,
+  ): void {
+    extractionMethod().subscribe({
+      next: response => {
+        var virtualMeters: Record<string, string> = {};
+        response.virtualMeters.forEach(virtualMeter => {
+          var ids = virtualMeter.id.split(":");
+          virtualMeters[virtualMeter.id] = ids[ids.length - 1];
+        });
+        this["optionsVirtualMeter"] = virtualMeters;
+        this["vMeters"] = response.virtualMeters;
+      },
+      error: error => {
+        console.log(error);
+      },
+    });
+  }
+
+  /**
+   * calls bws service to retrieve all physical meter information
+   */
+  extractPMeters(): void {
+    this.extractData(
+      () => this.bwsService.getPhysicalMeters(),
+      "meters",
+      "pMeters",
+    );
+  }
+
+  /**
+   * calls bws service to retrieve all virtual meter information
+   */
+  extractVMeters(): void {
+    this.extractVirtualMetersInRecord(() => this.bwsService.getVirtualMeters());
+  }
+
+  /**
+   * returns an VirtualMeter based on the given id
+   */
+  getVirtualMeter(virtualMeterId: string): VirtualMeter | undefined {
+    var vMeter = this.vMeters.filter(vMeter => vMeter.id === virtualMeterId);
+
+    console.log(vMeter);
+
+    if (vMeter.length === 1) {
+      return vMeter[0];
+    }
+
+    return undefined;
+  }
+
+  /**
+   * calls bws service to retrieve all algorithms
+   */
+  extractAlgorithms(): void {
+    this.extractAlgorithmsInRecord(() => this.bwsService.getAlgorithms());
+  }
+
+  /**
+   * returns an algorithm based on the given name
+   */
+  getAlgorithm(algorithmName: string): Algorithm | undefined {
+    var algorithm = this.algorithms.filter(
+      algorithm => algorithm.name === algorithmName,
+    );
+
+    if (algorithm.length === 1) {
+      return algorithm[0];
+    }
+
+    return undefined;
+  }
+
+  /**
+   * calls bws service to retrieve all trained models
+   */
+  extractModels(): void {
+    this.extractData(() => this.bwsService.getModels(), "MLModels", "models");
+  }
+
+  // ---------- Checkbox Functions ----------
+
+  toggleSelectedMeter(item: any, event: Event, selectedMeters: any): void {
+    const isChecked = (event.target as HTMLInputElement).checked;
+
+    if (isChecked) {
+      selectedMeters.push(item);
+    } else {
+      const index = selectedMeters.findIndex(
+        (meter: {id: any}) => meter.id === item.id,
+      );
+      if (index > -1) {
+        selectedMeters.splice(index, 1); // Remove the item if unchecked
+      }
+    }
+  }
```

Comment:
This feels a bit obscure to me, I think we can make this more easily using signals